### PR TITLE
LSTM Article added

### PR DESCRIPTION
**Pull request by nimbus98**
Added the LSTM Article and also corrected a line in Qlearning article that madhuparna requested to correct.

### CODE REVIEW

File: _posts/2018-12-13-LSTM.markdown [hunk 1]
```diff
@@ -0,0 +1,387 @@
+---
+layout: post
+title:  "Long Short Term Memory Networks"
+date:   2018-12-13 10:10:04 +0400
+categories: jekyll update
+author: Akash Nair
+---
+
+In this article, we will learn about a specific Recurrent Neural Network known as Long Short Term Memory Neural Network, or LSTMs in short. We will also be implementing a basic character to character RNN using LSTM using pyTorch.
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: should be 'basic character to character LSTM using pyTorch'

File: _posts/2018-12-13-LSTM.markdown [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+
+## What is a Recurrent Neural Network and Why LSTMs?
+
+
+**Recurrent Neural Networks** are those Neural Networks that we use to process information that require us to keep informed of previous information.
+In simpler words, when we want a model to perform a certain category of tasks like speech recognition, music generation, machine translation, sentiment classification, etc, all of which invloves keeping track of previous information (like keeping track of all the words while processing a sentence during machine translation), a normal Neural Network cannot do this, while a Recurrent Neural Network, which persists the same information having loops in them addresses this issue.
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: 'which persists the same information having loops in them addresses this issue.' this part isn't quite clear. Maybe something like 'which keeps a track of the past information in it's internal state'

File: _posts/2018-12-13-LSTM.markdown [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+|Recurrent Neural Network - Unrolled|
+|:-------------------------------------------------:|
+|![RNN](/assets/RNN.png){:height="75%" width="75%"}|
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: The image isn't a square, remove the height and width part

File: _posts/2018-12-13-LSTM.markdown [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+
+The above diagram is that of a basic Recurrent Neural Network, the chained structure is appropriate for sequence models and using this structure for Neural Networks have proven to work for the above mentioned applications. LSTMs come into play when a certain problem occurs within RNNs itself.
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: 'appropriate for modeling sequences'

File: _posts/2018-12-13-LSTM.markdown [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+This is where LSTMs come into play!
+
+## Long Short Term Memory Networks
+
+**Long Short Term Memory Networks** (usually just called `LSTMs`) are a _special kind_ of RNN, capable of learning long-term dependencies. They were introduced by Hochreiter & Schmidhuber (1997). They are explicitly designed to avoid the long-term dependency problem by remembering information for long periods of time, and this is possible by introducing 'memory cells' which keep track of these dependencies throughout the sequence.
+
+|Long Short Term Memory Network|
+|:-------------------------------------------------:|
+|![LSTM](/assets/LSTM.png){:height="75%" width="75%"}|
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: Remove height and width

File: _posts/2018-12-13-LSTM.markdown [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+
+This is how an LSTM looks like, it follows the same chain like structure like that of the RNNs, but it contains several added layers. It may seem difficult to process this as a whole, but we'll walk through this step by step.
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: Calling the gates 'layers' is quite misleading and confusing.

File: _posts/2018-12-13-LSTM.markdown [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+
+
+**Note**: The sigmoid function returns a value between 0 to 1 and this case, values very close to either 0 or 1 and hence is commonly used in our gates to make a particular decision.
+
+
+The following are layers involved in an LSTM cell:
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: Again, avoid using 'layers' for gates.

File: _posts/2018-12-13-LSTM.markdown [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+1. **Activation layer**: This layer consists of the activation values like the normal RNNs do
+2. **Memory Cell** or **Candidate layer**: This layer is involved in keeping track of dependencies
+3. **Update Gate**: Sigmoid function that decides whether or not the memory cell should keep track of the dependecy
+4. **Forget Gate**: Sigmoid function that decides whether or not the memory cell should leave or forget the dependency
+5. **Output Gate**: Sigmoid function that helps us filter what parts of the memory cell layer we want to pass into the output
+
+|LSTM Structure Inside|
+|:-------------------------------------------------:|
+|![LSTM2](/assets/LSTM2.png){:height="75%" width="75%"}|
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: remove width and height

File: _posts/2018-12-13-LSTM.markdown [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+
+$$\Gamma_{f} = \sigma(  W_{f}[ a^{<t-1>},x^{<t>} ] + b_f  )$$
+
+Here, the decision is made whether or not to keep track of the dependency with the help of the update and forget gates, which are sigmoid layers.
+
+$$\Gamma_{o} = \sigma(  W_{o}[ a^{<t-1>},x^{<t>} ] + b_o  )$$
+
+$$c^{<t>} = \Gamma_{u}*\hat{c}^{<t>} + \Gamma_{f}*c^{<t-1>}$$
+
+Here, we decide what exactly to update into the memory cell, which either retains the dependency from earlier or updates it to a new value based on the decision made by the update and forget gates. Hence,, the output will be filtered. This is done by running a sigmoid function layer to decide which parts of the memory cell we will send to the output and while the memory cells will be passed through tanh and then passed through the output gate to get only the filtered output. Here, our memory cells are updated appropriately.
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: double commas after hence

File: _posts/2018-12-13-LSTM.markdown [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+$$c^{<t>} = \Gamma_{u}*\hat{c}^{<t>} + \Gamma_{f}*c^{<t-1>}$$
+
+Here, we decide what exactly to update into the memory cell, which either retains the dependency from earlier or updates it to a new value based on the decision made by the update and forget gates. Hence,, the output will be filtered. This is done by running a sigmoid function layer to decide which parts of the memory cell we will send to the output and while the memory cells will be passed through tanh and then passed through the output gate to get only the filtered output. Here, our memory cells are updated appropriately. 
+
+$$a^{<t>} = \Gamma_{o}*tanh(c^{<t>})$$
+
+The activation layer is influenced by certain memory cell values decided upon by the output gate, and is appropriately updated and passed onto the next cell.
+
+
+The resultant $$\hat{y}^{<t>}$$ vector is obtained by passing the activation layer through a softmax function.
```
Comment (MJ10) [inline _posts/2018-12-13-LSTM.markdown]: I think you should mention that this step is dependent on what problem you're solving and isn't part of the general LSTM framework

**Review by MJ10 — CHANGES_REQUESTED** [review]
Great work on the article!
There are a few issues just resolve them and it should be done.

**Comment by nimbus98** [comment]
The corrections are made!
Thank you ^^

**Review by MJ10 — APPROVED** [review]
Looks good!

**Comment by mahirjain25** [comment]
Looks good to me  :+1:

**Pull request merged**